src/sage/algebras/lie_algebras/superliealgebra.py

```diff
@@ -0,0 +1,354 @@
+
+        spanning_set = [X.to_vector() for X in list(on_generators)]
+        im_gens = [Y.to_vector() for Y in list(on_generators.values())]
+
+        def solve_linear_system(A, b):
+            R = QQ
+            A_inv = A.solve_left(matrix.identity(A.ncols()))
+            M = A * A_inv
+            for Mi, bk in zip(M.rows(), b):
+                test_bk = sum((R(Mij) * bj for Mij,bj in zip(Mi,b)), cm.zero())
+                if test_bk != bk:
+                    raise ValueError("contradictory linear system")
+
+            return [sum((R(Aij) * bk for Aij,bk in zip(Ai,b)), cm.zero())
+                    for Ai in A_inv.rows()]
+
+        bracketlength = 1
+        n = 0
+        while True:
+            sm = m.submodule(spanning_set)
+            A = matrix(sm.base_ring(), [sm.coordinate_vector(X) for X in spanning_set])
+            try:
+                im_gens = solve_linear_system(A, im_gens)
+            except ValueError:
+                raise ValueError("this does not define a graded Lie algebra morphism")
+
+            spanning_set = list(sm.basis())
+            if n == len(spanning_set):
+
+                break
+
+            bracketlength += 1
+            n = len(spanning_set)
+            for i,j in combinations(range(n), 2):
+
+                Z = list(on_generators.keys())[i].bracket(list(on_generators.keys())[j])
+                imZ = list(on_generators.values())[i].bracket(list(on_generators.values())[j])
+                spanning_set.append(Z.to_vector())
+                im_gens.append(imZ.to_vector())
+
+        A = matrix(m.base_ring(), spanning_set)
+        im_gens = solve_linear_system(A, im_gens)
+
+        bh = lambda t: t
+        return cm.sum(bh(c) * im_gens[i] for i, c in (argument.to_vector()).items())
+
+
+    class Element(CombinatorialFreeModule.Element):
+         
+        #Returns the degree of the highest-degree monomial in any expression:
+        def degree(self):
+
+            l = [m for m in self]
+            monomials = [x[0] for x in l]
+            dictionary = self.parent().name_degree_map
+            degrees=self.parent()._degrees
+
+            values = [dictionary[key] for key in monomials if key in dictionary]
+            if values:
+                return max(values)
+            else:
+                return 0
+            
+        #Returns the degree of a homogeneous expression:
+        def homdegree(self):
+
+            l = [m for m in self]
+            monomials = [x[0] for x in l]
+            dictionary = self.parent().name_degree_map
+            degrees=self.parent()._degrees
+
+            values = [dictionary[key] for key in monomials if key in dictionary]
+            if values:
+                if len(values)!=1:
+                    return "Not homogeneous"
+                else:
+                    return values[0]
+            else:
+                return 0
+
+        def bracket(self, rt):
+            names=self.parent()._names
```

Comment:
```suggestion
            P = self.parent()
            names = P._names
```
Store the parent in a variable (it adds up all of these repeated (Python) function calls); similar changes below.